### PR TITLE
refactor(Script/Command): modify mount & dismount

### PR DESCRIPTION
**Pull request by Kitzunu**
Co-Authored-By: Rochet2 <rochet2@post.com>

Cherry-pick commit <https://github.com/TrinityCore/TrinityCore/commit/7c39123432c969173b168613d73b762eac67a303>

## Changes Proposed:
- Allow to use the command .mod mount #creatureDisplayID #speed
- #speed can be between 0.1 - 50
- Default #speed == 1 which means it is an optional parameter.
- .dismount will now correctly update your speed to normal running speed.
- Update command description

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/7374

## Tests Performed:

- Builds
- Tested in-game

## How to Test the Changes:

1. .mod mount 14340 0.1

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by LightDestory** [comment]
I am not an official tester, but as the issue's author I am gonna test anyway :)

Let's wait for the build

**Comment by Kitzunu** [comment]
Everyone can test PR's :D

**Comment by LightDestory** [comment]
Can I suggest a change? 
The old script was using a pre-defined speed value. We could use it again. So, to keep it short, we could make the second parameters optional using a default value.

**Comment by Kitzunu** [comment]
I guess we can default it to 1 which is 100% speed (walking speed)

**Comment by LightDestory** [comment]
Tested with 
`.modify mount 14340 20` and it works.
image

**Comment by Kitzunu** [comment]
I will update speed to default 1.f tomorrow or the next day.

Thanks for testing

File: src/server/scripts/Commands/cs_modify.cpp [hunk 1]
```diff
@@ -75,6 +75,16 @@ class modify_commandscript : public CommandScript
         return commandTable;
     }
 
+    template<typename... Args> static void NotifyModification(ChatHandler* handler, Unit* target, AcoreStrings resourceMessage, AcoreStrings resourceReportMessage, Args&&... args)
+    {
+        if (Player* player = target->ToPlayer())
+        {
+            handler->PSendSysMessage(resourceMessage, handler->GetNameLink(player).c_str(), args...);
+            if (handler->needReportToTarget(player))
+                ChatHandler(player->GetSession()).PSendSysMessage(resourceReportMessage, handler->GetNameLink().c_str(), std::forward<Args>(args)...);
```
Comment (Helias) [inline src/server/scripts/Commands/cs_modify.cpp:84]:
```suggestion
            if (handler->needReportToTarget(player))
            {
                ChatHandler(player->GetSession()).PSendSysMessage(resourceReportMessage, handler->GetNameLink().c_str(), std::forward<Args>(args)...);
            }
```

File: src/server/scripts/Commands/cs_modify.cpp [hunk 2]
```diff
@@ -454,6 +464,43 @@ class modify_commandscript : public CommandScript
         return false;
     }
 
+    static bool CheckModifySpeed(ChatHandler* handler, char const* args, Unit* target, float& speed, float minimumBound, float maximumBound, bool checkInFlight = true)
+    {
+        if (!*args)
+            return false;
```
Comment (Helias) [inline src/server/scripts/Commands/cs_modify.cpp:470]:
```suggestion
        if (!*args)
        {
            return false;
        }
```

File: src/server/scripts/Commands/cs_modify.cpp [hunk 3]
```diff
@@ -454,6 +464,43 @@ class modify_commandscript : public CommandScript
...
+                return false;
+
+            if (player->IsInFlight() && checkInFlight)
+            {
+                handler->PSendSysMessage(LANG_CHAR_IN_FLIGHT, handler->GetNameLink(player).c_str());
+                handler->SetSentErrorMessage(true);
+                return false;
+            }
+        }
+        return true;
```
Comment (Helias) [inline src/server/scripts/Commands/cs_modify.cpp:501]:
```suggestion
        }

        return true;
```

File: src/server/scripts/Commands/cs_modify.cpp [hunk 4]
```diff
@@ -454,6 +466,46 @@ class modify_commandscript : public CommandScript
...
     }
 
+    static bool CheckModifySpeed(ChatHandler* handler, char const* args, Unit* target, float& speed, float minimumBound, float maximumBound, bool checkInFlight = true)
+    {
+        if (!*args)
+        {
+            return false;
+        }
+
+        speed = (float)atof((char*)args);
```
Comment (Winfidonarleyan) [inline src/server/scripts/Commands/cs_modify.cpp:476]: Use StringTo

File: src/common/Common.h [hunk 5]
```diff
@@ -41,6 +41,9 @@
 
 #define MAX_NETCLIENT_PACKET_SIZE (32767 - 1)               // Client hardcap: int16 with trailing zero space otherwise crash on memory free
 
+inline unsigned long atoul(char const* str) { return strtoul(str, nullptr, 10); }
```
Comment (Winfidonarleyan) [inline src/common/Common.h:44]: No no no, please

Comment (Kitzunu) [reply src/common/Common.h:44]: What do you suggest instead? I didn't quite understand this part so I am more than happy to change and learn =)

Comment (Kitzunu) [reply src/common/Common.h:44]: Maybe I just convert the string to uint32 and not unsigned long (long)

Comment (Kitzunu) [reply src/common/Common.h:44]: Or I can use Acore::StringTo<uint32>? Will test that

File: src/server/scripts/Commands/cs_modify.cpp (hunk 4 above)
Comment (Kitzunu) [reply src/server/scripts/Commands/cs_modify.cpp:476]: changed to std::atof for now

**Comment by FrancescoBorzi** [comment]
what's the status of this?

**Comment by Kitzunu** [comment]
Waiting for @Winfidonarleyan to review

File: src/server/scripts/Commands/cs_modify.cpp [hunk 6]
```diff
@@ -454,6 +468,48 @@ class modify_commandscript : public CommandScript
...
     }
 
+    static bool CheckModifySpeed(ChatHandler* handler, char const* args, Unit* target, float& speed, float minimumBound, float maximumBound, bool checkInFlight = true)
+    {
+        if (!*args)
+        {
+            return false;
+        }
+
+        speed = double(std::atof(args));
```
Comment (Winfidonarleyan) [inline src/server/scripts/Commands/cs_modify.cpp:478]:
```suggestion
        speed = Acore::StringTo<float>(args).value();
```

Comment (Winfidonarleyan) [reply src/server/scripts/Commands/cs_modify.cpp:478]: Try this

Comment (Winfidonarleyan) [reply src/server/scripts/Commands/cs_modify.cpp:478]: See example https://github.com/azerothcore/azerothcore-wotlk/blob/888549bcdc2622a5f75cd1765e3915e17c9e8ead/src/server/scripts/Commands/cs_account.cpp#L93

Comment (Kitzunu) [reply src/server/scripts/Commands/cs_modify.cpp:478]: maybe .value will fix it, but just StringTo will fail. Will commit and let CI test

File: src/server/scripts/Commands/cs_modify.cpp [hunk 7]
```diff
@@ -725,229 +781,32 @@ class modify_commandscript : public CommandScript
...
+            return false;
         }
 
-        Player* target = handler->getSelectedPlayer();
+        if (!speed_cstr)
+        {
+            speed_cstr = "1";
+        }
+
+        uint32 mount = uint32(std::stoul(mount_cstr));
```
Comment (Winfidonarleyan) [inline src/server/scripts/Commands/cs_modify.cpp:801]: Acore string to

Comment (Kitzunu) [reply src/server/scripts/Commands/cs_modify.cpp:801]: Kek missed it

**Pull request merged**